File: src/ardldml/folds.py
"""
h-block cross-fitting for time series.

Cross-fitting removes the bias that arises when the same observations are used
both to estimate a nuisance function and to evaluate the score built from it.
In cross-sectional DML the folds are random; in a time series they cannot be,
because randomly held-out observations are adjacent in time to their own
training data and the first-stage error is then correlated with the evaluation
innovations.

DML-Bounds uses the *h-block* scheme: the sample is cut into ``K`` contiguous
chronological blocks, and each block is evaluated using a model trained on the
other blocks *minus an ``h``-observation buffer* on either side of the
evaluation window. The buffer is what buys the decoupling -- Lemma 2 of the
paper turns Neyman orthogonality from an assumption into a consequence of the
construction, provided :math:`h \\to \\infty` and :math:`h = o(T)`.

The cost is sample. The buffer discards at most :math:`h(K-1)` observations
from the training sets, which is second-order asymptotically (Remark 7 notes
:math:`h(K-1)/T \\to 0`) but very much first-order at the sample sizes applied
ARDL work uses. :func:`sample_use` reports exactly how much is lost, so the
trade-off is visible rather than implicit.

Choosing ``h`` and ``K``
------------------------
``h`` should cover the memory of the process: with monthly data and an ARDL of
order 4, ``h`` of 6-12 is reasonable. ``K`` trades bias against the size of
each training set; ``K = 5`` is the default here and ``K = 6`` is used in the
paper's applications. Larger ``K`` means more, smaller, training sets and more
buffer loss in total.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Tuple

import numpy as np
import pandas as pd

__all__ = ["BlockStructure", "hblock_folds", "sample_use", "sample_use_table"]


@dataclass
class BlockStructure:
    """
    The evaluation/training partition produced by :func:`hblock_folds`.

    Attributes
    ----------
    n : int
        Number of observations.
    n_blocks : int
        Number of chronological blocks ``K``.
    buffer : int
        Buffer width ``h`` in observations.
    eval_blocks : list of numpy.ndarray
        Index arrays of the evaluation windows. These partition ``range(n)``.
    train_blocks : list of numpy.ndarray
        Index arrays of the corresponding training samples, each excluding its
        own evaluation window and the ``h`` observations either side of it.
    """

    n: int
    n_blocks: int
    buffer: int
    eval_blocks: List[np.ndarray] = field(default_factory=list)
    train_blocks: List[np.ndarray] = field(default_factory=list)

    def __len__(self) -> int:
        return self.n_blocks

    def __iter__(self):
        return iter(zip(self.train_blocks, self.eval_blocks))

    def to_frame(self) -> pd.DataFrame:
        """One row per block, with its extent and training size."""
        rows = []
        for b, (tr, ev) in enumerate(zip(self.train_blocks, self.eval_blocks), start=1):
            rows.append(
                {
                    "block": b,
                    "eval_start": int(ev[0]),
                    "eval_end": int(ev[-1]),
                    "n_eval": int(ev.size),
                    "n_train": int(tr.size),
                    "share_train": tr.size / self.n,
                }
            )
        return pd.DataFrame(rows).set_index("block")

    def __repr__(self) -> str:  # pragma: no cover - cosmetic
        return (
            f"<BlockStructure n={self.n} K={self.n_blocks} h={self.buffer} "
            f"mean train share={np.mean([t.size for t in self.train_blocks]) / self.n:.2f}>"
        )


def hblock_folds(n: int, n_blocks: int = 5, buffer: int = 0) -> BlockStructure:
    """
    Build contiguous evaluation blocks with buffered training sets.

    Parameters
    ----------
    n : int
        Number of observations.
    n_blocks : int
        Number of chronological blocks ``K``. Must be at least 2.
    buffer : int
        Buffer width ``h``. Observations within ``h`` of an evaluation window
        are dropped from that window's training set.

    Returns
    -------
    BlockStructure

    Raises
    ------
    ValueError
        If the buffer is so wide that some block has an empty training set,
        which happens when ``K`` is small and ``h`` is large relative to ``n``.

    Examples
    --------
    >>> bs = hblock_folds(100, n_blocks=4, buffer=5)
    >>> bs.to_frame()["n_train"].tolist()
    [70, 65, 65, 70]
    """
    if n_blocks < 2:
        raise ValueError(f"n_blocks must be at least 2; got {n_blocks!r}")
    if buffer < 0:
        raise ValueError(f"buffer must be non-negative; got {buffer!r}")
    if n_blocks > n:
        raise ValueError(f"n_blocks={n_blocks} exceeds n={n}")

    edges = np.linspace(0, n, n_blocks + 1).astype(int)
    struct = BlockStructure(n=int(n), n_blocks=int(n_blocks), buffer=int(buffer))
    all_idx = np.arange(n)

    for b in range(n_blocks):
        lo, hi = edges[b], edges[b + 1]
        ev = all_idx[lo:hi]
        keep = np.ones(n, dtype=bool)
        keep[max(0, lo - buffer):min(n, hi + buffer)] = False
        tr = all_idx[keep]
        if tr.size == 0:
            raise ValueError(
                f"block {b + 1} has an empty training set: buffer={buffer} is too wide "
                f"for n={n} with n_blocks={n_blocks}"
            )
        struct.eval_blocks.append(ev)
        struct.train_blocks.append(tr)
    return struct


def sample_use(n: int, n_blocks: int = 5, buffer: int = 0) -> float:
    """
    Average share of the sample available for training, across blocks.

    This is the quantity that makes the cost of buffering concrete. With no
    buffer it is ``1 - 1/K``; each unit of ``h`` removes roughly ``2h/n`` more.
    """
    bs = hblock_folds(n, n_blocks=n_blocks, buffer=buffer)
    return float(np.mean([t.size for t in bs.train_blocks]) / n)


def sample_use_table(
    n: int,
    n_blocks: Tuple[int, ...] = (4, 5, 6, 8),
    buffers: Tuple[int, ...] = (0, 2, 5, 10),
) -> pd.DataFrame:
    """
    Training-sample share for a grid of ``K`` and ``h``.

    Useful before committing to a configuration on a short sample: it shows at
    a glance when the buffer starts to cost more than the cross-fitting buys.

    Returns
    -------
    pandas.DataFrame
        Rows indexed by ``K``, columns by ``h``, entries the average training
        share. ``NaN`` marks configurations that are infeasible.
    """
    out = pd.DataFrame(index=pd.Index(n_blocks, name="K"), columns=list(buffers), dtype=float)
    out.columns.name = "h"
    for k in n_blocks:
        for h in buffers:
            try:
                out.loc[k, h] = sample_use(n, n_blocks=k, buffer=h)
            except ValueError:
                out.loc[k, h] = np.nan
    return out

File: r/ardldml/R/folds.R
#' h-Block Cross-Fitting Folds
#'
#' Cuts the sample into \code{n_blocks} contiguous chronological blocks and
#' builds, for each, a training set that excludes the block itself and the
#' \code{buffer} observations either side of it.
#'
#' In cross-sectional double machine learning the folds are random. In a time
#' series they cannot be: randomly held-out observations are adjacent in time to
#' their own training data, so the first-stage error stays correlated with the
#' evaluation innovations. The buffer is what buys the decoupling, and it costs
#' sample -- at most \eqn{h(K-1)} observations across the training sets. Use
#' \code{\link{sample_use_table}} to see the cost before choosing a
#' configuration.
#'
#' @param n Number of observations.
#' @param n_blocks Number of chronological blocks \eqn{K}; at least 2.
#' @param buffer Buffer width \eqn{h} in observations.
#'
#' @return An object of class \code{"ardl_folds"}: a list with \code{n},
#'   \code{n_blocks}, \code{buffer}, and lists \code{eval} and \code{train} of
#'   integer index vectors.
#'
#' @examples
#' f <- hblock_folds(100, n_blocks = 4, buffer = 5)
#' sapply(f$train, length)
#'
#' @seealso \code{\link{sample_use}}, \code{\link{sample_use_table}}
#' @export
hblock_folds <- function(n, n_blocks = 5L, buffer = 0L) {
  n <- as.integer(n)
  n_blocks <- as.integer(n_blocks)
  buffer <- as.integer(buffer)
  if (n_blocks < 2L) stop("n_blocks must be at least 2; got ", n_blocks)
  if (buffer < 0L) stop("buffer must be non-negative; got ", buffer)
  if (n_blocks > n) stop("n_blocks = ", n_blocks, " exceeds n = ", n)

  edges <- as.integer(seq(0, n, length.out = n_blocks + 1L))
  eval_blocks <- vector("list", n_blocks)
  train_blocks <- vector("list", n_blocks)

  for (b in seq_len(n_blocks)) {
    lo <- edges[b]
    hi <- edges[b + 1L]
    ev <- seq.int(lo + 1L, hi)
    keep <- rep(TRUE, n)
    a <- max(0L, lo - buffer)
    z <- min(n, hi + buffer)
    if (z > a) keep[seq.int(a + 1L, z)] <- FALSE
    tr <- which(keep)
    if (length(tr) == 0L) {
      stop("block ", b, " has an empty training set: buffer = ", buffer,
           " is too wide for n = ", n, " with n_blocks = ", n_blocks)
    }
    eval_blocks[[b]] <- ev
    train_blocks[[b]] <- tr
  }

  structure(
    list(n = n, n_blocks = n_blocks, buffer = buffer,
         eval = eval_blocks, train = train_blocks),
    class = "ardl_folds"
  )
}

#' @rdname hblock_folds
#' @param x An \code{"ardl_folds"} object.
#' @param ... Ignored.
#' @export
as.data.frame.ardl_folds <- function(x, ...) {
  data.frame(
    block = seq_len(x$n_blocks),
    eval_start = vapply(x$eval, function(e) e[1L], numeric(1)),
    eval_end = vapply(x$eval, function(e) e[length(e)], numeric(1)),
    n_eval = vapply(x$eval, length, numeric(1)),
    n_train = vapply(x$train, length, numeric(1)),
    share_train = vapply(x$train, length, numeric(1)) / x$n
  )
}

#' @rdname hblock_folds
#' @export
print.ardl_folds <- function(x, ...) {
  cat(sprintf("h-block folds: n = %d, K = %d, h = %d, mean training share = %.2f\n",
              x$n, x$n_blocks, x$buffer,
              mean(vapply(x$train, length, numeric(1))) / x$n))
  print(as.data.frame(x), row.names = FALSE)
  invisible(x)
}

#' Share of the Sample Available for Training
#'
#' Average, across blocks, of the training-set size as a fraction of \code{n}.
#' With no buffer this is \eqn{1 - 1/K}; each unit of \eqn{h} removes roughly
#' \eqn{2h/n} more. This is the quantity that makes the cost of buffering
#' concrete.
#'
#' @inheritParams hblock_folds
#' @return A single number between 0 and 1.
#' @examples
#' sample_use(108, n_blocks = 5, buffer = 6)
#' @export
sample_use <- function(n, n_blocks = 5L, buffer = 0L) {
  f <- hblock_folds(n, n_blocks = n_blocks, buffer = buffer)
  mean(vapply(f$train, length, numeric(1))) / f$n
}

#' Training Share Across a Grid of K and h
#'
#' Tabulates \code{\link{sample_use}} over a grid, so the point at which the
#' buffer costs more than the cross-fitting buys is visible before a
#' configuration is committed to. Infeasible cells are \code{NA}.
#'
#' @param n Number of observations.
#' @param n_blocks Integer vector of block counts \eqn{K}.
#' @param buffers Integer vector of buffer widths \eqn{h}.
#' @return A numeric matrix, rows named by \eqn{K} and columns by \eqn{h}.
#' @examples
#' sample_use_table(108)
#' @export
sample_use_table <- function(n, n_blocks = c(4L, 5L, 6L, 8L),
                             buffers = c(0L, 2L, 5L, 10L)) {
  out <- matrix(NA_real_, nrow = length(n_blocks), ncol = length(buffers),
                dimnames = list(paste0("K=", n_blocks), paste0("h=", buffers)))
  for (i in seq_along(n_blocks)) {
    for (j in seq_along(buffers)) {
      out[i, j] <- tryCatch(
        sample_use(n, n_blocks = n_blocks[i], buffer = buffers[j]),
        error = function(e) NA_real_
      )
    }
  }
  out
}
